server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/rust/RustWorkspaceResolver.kt

```diff
@@ -0,0 +1,278 @@
+package org.jetbrains.bsp.bazel.server.sync.languages.rust
+
+import ch.epfl.scala.bsp4j.RustDepKindInfo
+import ch.epfl.scala.bsp4j.RustDependency
+import ch.epfl.scala.bsp4j.RustEnvData
+import ch.epfl.scala.bsp4j.RustFeature
+import ch.epfl.scala.bsp4j.RustPackage
+import ch.epfl.scala.bsp4j.RustProcMacroArtifact
+import ch.epfl.scala.bsp4j.RustRawDependency
+import ch.epfl.scala.bsp4j.RustTarget
+import org.jetbrains.bsp.bazel.server.sync.BazelPathsResolver
+import org.jetbrains.bsp.bazel.server.sync.model.Label
+import org.jetbrains.bsp.bazel.server.sync.model.Module
+
+private data class BazelPackageTargetInfo(
+    val packageName: String,
+    val targetName: String
+)
+
+private data class RustTargetModule(
+    val module: Module,
+    val rustModule: RustModule
+)
+
+class RustWorkspaceResolver(val bazelPathsResolver: BazelPathsResolver) {
+    fun rustPackages(rustBspTargets: List<Module>): List<RustPackage> = rustBspTargets
+        .groupBy { resolvePackage(it).packageName }
+        .mapNotNull(::resolveSinglePackage)
+
+    // We need to resolve all dependencies are provide a list of new bazel targets (deps) to be transformed into packages.
+    fun rustDependencies(
+        rustPackages: List<RustPackage>,
+        rustBspTargets: List<Module>
+    ): Pair<List<RustDependency>, List<RustRawDependency>> {
+
+        val associatedBspTargets = groupBspTargetsByPackage(rustBspTargets, rustPackages)
+        val associatedRawBspTargets = groupBspRawTargetsByPackage(rustBspTargets, rustPackages)
+
+        val rustDependencies = resolveDependencies(associatedBspTargets)
+        val rustRawDependencies = resolveRawDependencies(associatedRawBspTargets)
+
+        return Pair(rustDependencies, rustRawDependencies)
+    }
+
+    private fun resolvePackage(rustTarget: Module): BazelPackageTargetInfo =
+        resolvePackage(rustTarget.label)
+
+    private fun resolvePackage(label: Label): BazelPackageTargetInfo {
+        val labelVal = label.value
+        val (packageName, targetName) = labelVal.split(":", limit = 2)
+        return BazelPackageTargetInfo(packageName, targetName)
+    }
+
+    // We need to merge targets with the same crate root.
+    private fun mergeTargets(rustTargets: List<RustTargetModule>): RustTargetModule {
+        require(rustTargets.isNotEmpty()) { "Cannot merge an empty list of targets." }
+        if (rustTargets.size == 1) {
```

Comment:
isn't this case covered by the `maxByOrNull` as well?

Reply:
Yep, it is